Fix PHP notice when we don't have credit card params

mattwire opened the pull request:
Overview
----------------------------------------
When using a credit card payment processor that doesn't submit card details to CiviCRM (eg. Stripe) we get PHP notices on backend submit credit card because it always tries to assign credit card params to the form.

Before
----------------------------------------
PHP notices:
![image](https://user-images.githubusercontent.com/2052161/74349701-84720500-4dac-11ea-8bb5-9d6435329352.png)


After
----------------------------------------
No PHP notices

Technical Details
----------------------------------------
Array indexes are being accessed without checking they exist. We already have a check for one of them a couple of lines above so move them there.

Comments
----------------------------------------

eileenmcnaughton commented on CRM/Contribute/Form/AbstractEditPayment.php:
```diff
@@ -562,10 +562,10 @@ protected function beginPostProcess() {
       if (in_array('credit_card_exp_date', array_keys($this->_params))) {
         $this->_params['year'] = CRM_Core_Payment_Form::getCreditCardExpirationYear($this->_params);
         $this->_params['month'] = CRM_Core_Payment_Form::getCreditCardExpirationMonth($this->_params);
+        $this->assign('credit_card_exp_date', CRM_Utils_Date::mysqlToIso(CRM_Utils_Date::format($this->_params['credit_card_exp_date'])));
```
So in general I prefer to ensure 'something' is assigned to the template - even if blank - this generally reduces leakage. In this case it's a form function so it's not clear leakage could occur. So I'm a soft preference for changing that.

I'd also prefer not to have params other than credit_card_exp_date in the if for that field. I know you can argue that there wouldn't be a card number with no date & I can't think of a case where that would be false but we are still using one field as a stand in for a greater meaning - I think

```
  $this->assign('credit_card_type', CRM_Utils_Array::value('credit_card_type', $this->_params));
```

per below is better practice

mattwire replied:
Yes. I messed this up - have pushed an updated version of `assignCreditCardDetails` which always assigns.

mattwire commented:
@eileenmcnaughton Ok this has ballooned a little bit after I addressed your comments. We actually have 7 places in the code where "credit card params" are formatted/assigned. By adding a new function `assignCreditCardParams` and modifying the existing `formatCreditCardParams` we can replace all these with the two functions (which also check if array indexes exist before using so we eliminate PHP notices due to missing credit card fields.

eileenmcnaughton commented:
@mattwire OK I'll double check - not assigning if empty means we are still susceptible to leakage - ie. this is better without the if
```
    if (!empty($params['credit_card_type'])) {
      $form->assign('credit_card_type', CRM_Utils_Array::value('credit_card_type', $params));
    }
```

Although that code may not be leakage prone (or at least not at the moment)

mattwire commented:
@eileenmcnaughton Ready for review now following your comments

mattwire commented:
test this please

eileenmcnaughton commented:
test this please

eileenmcnaughton commented:
@mattwire I started looking at this but still have some reservations in the portion that I looked at 

1) $form & $params being passed as reference
2) the function is called - assign but  is actually retrieving 
3) this line just looks like the sort of civihell approach where we have a function do something different if we pass an extra param. 
```  $tplParams = array_merge($tplParams, CRM_Contribute_Form_AbstractEditPayment::assignCreditCardDetails($this, $processorParams, FALSE));```

- perhaps we need  just a parseCardTemplateParams function & then  at the CRM_Form level we could have a function assignVariablesFromParams or similar that takes an array of keys + params & does much like the assign part of your function.

I pulled this out to get it merged - https://github.com/civicrm/civicrm-core/pull/16562 - since it  took a bit  of spotting but makes sense

mattwire commented:
Ok @eileenmcnaughton I'm trying to consolidate code without too much duplication here. How about if `formatCreditCardDetails` returns an array of template params (it currently returns nothing). Then we add a function on `CRM_Core_Form` that accepts an array of key=>value params and assigns them to the form?

eileenmcnaughton commented:
how about calling it getFormattedCreditCardParams or parseFormattedCreditCardParams

I'm imagining the assign function something like

```
protected function assignVariables($values, $keys) {
   foreach ($keys as $key)  {
     $this-assign($key, $values[$key] ?? NULL);
   }
}
```

mattwire commented:
Thanks @eileenmcnaughton will update the PR soon.

mattwire commented:
@eileenmcnaughton I've implemented your suggestions - now ready for review again

jaapjansma commented:
Betty and I are reviewing PR's and we came across this PR. However we miss a clear step by step instruction on how we could reproduce this as a user. We also think we need the strip payment processor extension to review this but not sure.

jaapjansma commented:
@mattwire this PR has conflicting files can you solve.

mattwire commented:
Hey @jaapjansma Thankyou. I have resolved and rebased.

eileenmcnaughton commented:
@mattwire just pinged me to look at this again - it's mostly a refactor & I'm comfortable merging it now - although I think the function would have been clearer if renamed per comment

mattwire commented:
Thankyou @eileenmcnaughton

eileenmcnaughton commented:
See https://github.com/civicrm/civicrm-core/pull/17584 for fix to regression from this PR
